nntrainer/include/databuffer_v2.h

```diff
@@ -0,0 +1,325 @@
+};
+
+/**
+ * @brief   Dataset generator callback type declaration
+ */
+typedef std::function<std::remove_pointer<ml_train_datagen_cb>::type>
+  datagen_cb;
+
+/**
+ * @class   DataBuffer Data Buffers
+ * @brief   Data Buffer for read and manage data
+ */
+class DataBuffer_v2 {
+public:
+  /**
+   * @brief     Constructor
+   */
+  DataBuffer_v2() :
+    type(DataBufferType::DATA_BUFFER_UNKNOWN),
+    buffer_len(1),
+    avail_buffer_idx(0),
+    total_data_entries(0),
+    batch_size(1),
+    num_threads(1),
+    generator(nullptr),
+    gen_user_data(nullptr),
+    thread_state(ThreadStates::THREAD_NULL) {
+    buffer.clear();
+    batched_buffer.clear();
+    label_size.resize(1, 0);
+    input_size.resize(1, 0);
+  }
+
+  /**
+   * @brief     Destructor
+   */
+  ~DataBuffer_v2() { stop(); }
+
+  /**
+   * @brief     Initialize Buffer with set properties
+   * @throws std::invalid_argument
+   * @throws std::runtime_error
+   */
+  void init();
+
+  /**
+   * @brief     start the thread for collection the data
+   * @throws std::runtime_error
+   */
+  void start();
+
+  /**
+   * @brief     function for thread to stop collection of the data
+   * @throws std::runtime_error
+   */
+  void stop();
+
+  /**
+   * @brief     Get data from buffer to the passed input and label vectors
+   * @param[in] inputs list of input tensors
+   * @param[in] inputs list of label tensors
+   * @retval false if end of epoch, else true
+   */
+  int getdata(std::vector<sharedtensor> &inputs,
+              std::vector<sharedtensor> &labels);
+
+  /**
+   * @brief     set the number of inputs (defaults to 1)
+   * @param[in] num_inputs number of inputs
+   * @throws std::invalid_argument
+   */
+  void setNumInputs(const unsigned int num_inputs = 1) {
+    if (num_inputs == 0)
+      throw std::invalid_argument("Number of inputs must be at least 1");
+
+    if (num_inputs != input_size.size()) {
+      input_size.resize(num_inputs, 0);
+    }
+  }
+
+  /**
+   * @brief     sets the number of labels (defaults to 1)
+   * @param[in] num_labels number of labels
+   * @throws std::invalid_argument
+   */
+  void setNumLabels(const unsigned int num_labels = 1) {
+    if (num_labels == 0)
+      throw std::invalid_argument("Number of labels must be at least 1");
+
+    if (num_labels != label_size.size()) {
+      label_size.resize(num_labels, 0);
+    }
+  }
+
+  /**
+   * @brief     set the size of the label data
+   * @param[in] bytes size in bytes
+   * @param[in] idx index of the label
+   * @throws std::invalid_argument
+   */
+  void setLabelSize(const size_t bytes, const unsigned int idx = 0) {
+    if (bytes == 0)
+      throw std::invalid_argument("Label size should be more than 0");
+
+    if (idx >= label_size.size()) {
+      if (label_size.size() > 1)
+        throw std::invalid_argument(
+          "Index exceeds the total size set for the label");
+      label_size.resize(idx + 1);
+    }
```

Comment:
Is this intended that resize only `if labe_size.size() == 1`?

Reply:
yes. The default label size is 1. If the caller does not edit the label size and tries to directly set a value outside this range, then it is allowed (determining size as caller sets the value). However, if the caller sets the label size to be higher, and then tries to set size outside the range, that's an error.

Reply:
Acked